### PR TITLE
feat: add http logging for c#

### PR DESCRIPTION
**Pull request by JenniferMah**
# Fixes #

https://issues.corp.twilio.com/browse/DI-945

### CODE REVIEW

File: src/Twilio/Clients/TwilioRestClient.cs [hunk 1]
```diff
@@ -70,6 +78,10 @@ public class TwilioRestClient : ITwilioRestClient
 
             Region = region;
             Edge = edge;
+            LogLevel = logLevel;
```
Comment (childish-sambino) [inline src/Twilio/Clients/TwilioRestClient.cs:81]: Simpler and provides precedence to the passed in log level:
```
LogLevel = logLevel ?? Environment.GetEnvironmentVariable("TWILIO_LOG_LEVEL");
```

File: src/Twilio/Clients/TwilioRestClient.cs [hunk 2]
```diff
@@ -92,6 +107,11 @@ public Response Request(Request request)
             try
             {
                 response = HttpClient.MakeRequest(request);
+                if(LogLevel == "debug")
```
Comment (childish-sambino) [inline src/Twilio/Clients/TwilioRestClient.cs:110]: nit: spacing

File: src/Twilio/Clients/TwilioRestClient.cs [hunk 3]
```diff
@@ -92,6 +107,11 @@ public Response Request(Request request)
...
+                {
+                    Console.WriteLine("response.status: " + response.StatusCode);
+                    Console.WriteLine("response.headers: " + "{"+ response.Headers + "}");
```
Comment (childish-sambino) [inline src/Twilio/Clients/TwilioRestClient.cs:113]: Why is this one specifically wrapped in braces?

File: src/Twilio/Clients/TwilioRestClient.cs [hunk 4]
```diff
@@ -92,6 +107,11 @@ public Response Request(Request request)
             try
             {
                 response = HttpClient.MakeRequest(request);
+                if(LogLevel == "debug")
+                {
+                    Console.WriteLine("response.status: " + response.StatusCode);
```
Comment (childish-sambino) [inline src/Twilio/Clients/TwilioRestClient.cs:112]: Is the console the best place to log?

File: src/Twilio/Clients/TwilioRestClient.cs [hunk 5]
```diff
@@ -220,5 +240,34 @@ public static void ValidateSslCertificate(HttpClient client)
                 );
             }
         }
+
+        /// <summary>
+        /// Format request information when LogLevel is set to debug
+        /// </summary>
+        ///
+        /// <param name="request">HTTP request</param>
+        private static void logRequest(Request request){
```
Comment (childish-sambino) [inline src/Twilio/Clients/TwilioRestClient.cs:249]: nit: `LogRequest`

File: src/Twilio/Clients/TwilioRestClient.cs [hunk 6]
```diff
@@ -220,5 +240,34 @@ public static void ValidateSslCertificate(HttpClient client)
...
+            Console.WriteLine("-- BEGIN Twilio API Request --");
+            Console.Write(request.Method + " ");
```
Comment (childish-sambino) [inline src/Twilio/Clients/TwilioRestClient.cs:251]: Why a trailing space?

File: src/Twilio/Clients/TwilioRestClient.cs [hunk 7]
```diff
@@ -220,5 +240,34 @@ public static void ValidateSslCertificate(HttpClient client)
...
+            Console.WriteLine(request.Uri);
```
Comment (childish-sambino) [inline src/Twilio/Clients/TwilioRestClient.cs:252]: We should be consistent about prefixing each log line with what we're logging.

File: src/Twilio/Clients/TwilioRestClient.cs [hunk 8]
```diff
@@ -53,13 +58,16 @@ public class TwilioRestClient : ITwilioRestClient
...
+
         public TwilioRestClient(
             string username,
             string password,
             string accountSid = null,
             string region = null,
             HttpClient httpClient = null,
-            string edge = null
+            string edge = null,
+            string logLevel = null
```
Comment (childish-sambino) [inline src/Twilio/Clients/TwilioRestClient.cs:70]: Should limit the number of constructor parameters. I think it's fine just to use a setter.

File: src/Twilio/Clients/TwilioRestClient.cs (hunk 3 above)
Comment (JenniferMah) [reply src/Twilio/Clients/TwilioRestClient.cs:113]: Was originally for readability because each header is logged onto a new line but the Node logging has the same format with just a label so I will remove the braces.

File: src/Twilio/Clients/TwilioRestClient.cs (hunk 6 above)
Comment (JenniferMah) [reply src/Twilio/Clients/TwilioRestClient.cs:251]: The method and URI were originally logged on the same line with no delineation between them but I decided to move them to two separate labeled lines.

File: src/Twilio/Clients/TwilioRestClient.cs [hunk 9]
```diff
@@ -220,5 +237,35 @@ public static void ValidateSslCertificate(HttpClient client)
...
+            if (request.QueryParams != null)
+            {
+                request.QueryParams.ForEach(kpv => Console.WriteLine(kpv.Key+ ":"+ kpv.Value));
+            }  
+
+            if (request.HeaderParams != null){
+                for(int i=0; i<request.HeaderParams.Count; i++)
+                {
+                    var lowercaseHeader = request.HeaderParams[i].Key.ToLower();
+                    if(lowercaseHeader.Contains("authorization") == false)
```
Comment (childish-sambino) [inline src/Twilio/Clients/TwilioRestClient.cs:261]: nit: more spacing issues. Use VS Code formatter?

File: src/Twilio/Clients/TwilioRestClient.cs [hunk 10]
```diff
@@ -70,6 +76,9 @@ public class TwilioRestClient : ITwilioRestClient
 
             Region = region;
             Edge = edge;
+
+            _logLevel = TwilioClient.logLevel ?? Environment.GetEnvironmentVariable("TWILIO_LOG_LEVEL");
```
[the remark this answers is not in the document]
Comment (JenniferMah) [reply src/Twilio/Clients/TwilioRestClient.cs:80]: Could I potentially use `LogLevel = logLevel ?? Environment.GetEnvironmentVariable("TWILIO_LOG_LEVEL")` after the rest client is built in twilio.cs? Then there is no need for a setter in the constructor?

[the remark this answers is not in the document]
Comment (childish-sambino) [reply src/Twilio/Clients/TwilioRestClient.cs:80]: Problem then is that anyone wanting to construct `TwilioRestClient` outside of the static singleton doesn't get to automatically use the env var.

File: src/Twilio/Clients/TwilioRestClient.cs (hunk 4 above)
Comment (JenniferMah) [reply src/Twilio/Clients/TwilioRestClient.cs:104]: After some investigation I think the console is the best place for now. If it becomes an issue later on then we can look into using a 3rd party logging library or other suggested solutions.

File: src/Twilio/Clients/TwilioRestClient.cs [hunk 11]
```diff
@@ -220,5 +232,36 @@ public static void ValidateSslCertificate(HttpClient client)
...
+        /// <param name="request">HTTP request</param>
+        private static void LogRequest(Request request)
+        {
+            Console.WriteLine("-- BEGIN Twilio API Request --");
+            Console.WriteLine("request.method: " + request.Method);
+            Console.WriteLine("request.URI: " + request.Uri);
+
+            if (request.QueryParams != null)
+            {
+                request.QueryParams.ForEach(kpv => Console.WriteLine(kpv.Key + ":" + kpv.Value));
```
Comment (eshanholtz) [inline src/Twilio/Clients/TwilioRestClient.cs:249]: what does kpv stand for?

Comment (JenniferMah) [reply src/Twilio/Clients/TwilioRestClient.cs:249]: It stands for the key pair value of each query param but if there is a different reference that makes this more readable I can change it.

Comment (eshanholtz) [reply src/Twilio/Clients/TwilioRestClient.cs:249]: Ah I usually use the term _key value pair (kvp)_, so the ordering threw me off. Mmmm I'd lean towards a more descriptive variable name. Usually if I can intuit what it represents it's good, but in this case I had to ask. Maybe just `param` => `param.Key` and `param.Value`?

**Comment by eshanholtz** [comment]
Can we add tests for this?

File: src/Twilio/Http/Request.cs [hunk 12]
```diff
@@ -24,7 +24,7 @@ public class Request
         /// </summary>
         public HttpMethod Method { get; }
 
-        private Uri Uri;
+        public Uri Uri;
```
Comment (eshanholtz) [inline src/Twilio/Http/Request.cs:27]: why this change?

File: test/Twilio.Test/Clients/TwilioRestClientTest.cs [hunk 13]
```diff
@@ -113,5 +113,15 @@ public void TestRedirectResponse()
...
             twilioClient.Request(request);
         }
+
+        [Test]
+        public void TestActivatingDebugLogging()
+        {
+            client.MakeRequest(Arg.Any<Request>()).Returns(new Response(HttpStatusCode.OK, "OK"));
+            Request request = new Request(HttpMethod.Get, "https://www.contoso.com");
+            TwilioRestClient twilioClient = new TwilioRestClient("foo", "bar", null, null, client);
+            twilioClient.LogLevel = "debug";
```
Comment (eshanholtz) [inline test/Twilio.Test/Clients/TwilioRestClientTest.cs:123]: Why are there no asserts here?

File: src/Twilio/Http/Request.cs (hunk 12 above)
Comment (JenniferMah) [reply src/Twilio/Http/Request.cs:27]: In order to log the request Uri in the _TwilioRestClient_ the Uri needs to be public.

Comment (eshanholtz) [reply src/Twilio/Http/Request.cs:27]: Then you should add the `{ get; private set; }` [like QueryParams](https://github.com/twilio/twilio-csharp/pull/545/files#diff-892488c88ebf07d450c8a3a13ccafcc08dfe907d422c5e9001f6aedd21b78718R52)

File: test/Twilio.Test/Clients/TwilioRestClientTest.cs (hunk 13 above)
Comment (JenniferMah) [reply test/Twilio.Test/Clients/TwilioRestClientTest.cs:123]: This test sets LogLevel on the _TwilioRestClient_ as debug but the only outcome is the logging output. What should I be asserting on?

Comment (eshanholtz) [reply test/Twilio.Test/Clients/TwilioRestClientTest.cs:123]: You should make a request and assert that the logging output exists and/or contains the "expected" value/content

File: test/Twilio.Test/Clients/TwilioRestClientTest.cs [hunk 14]
```diff
@@ -7,6 +7,8 @@
 using Twilio.Clients;
 using Twilio.Exceptions;
 using Twilio.Http;
+using Twilio.Rest.Api.V2010.Account;
```
Comment (eshanholtz) [inline test/Twilio.Test/Clients/TwilioRestClientTest.cs:10]: Is this necessary?

Comment (JenniferMah) [reply test/Twilio.Test/Clients/TwilioRestClientTest.cs]: ops! Nope I will fix that right now.

**Review by eshanholtz — APPROVED** [review]
noice

File: src/Twilio/Twilio.cs [hunk 15]
```diff
@@ -126,6 +127,20 @@ public static void SetEdge(string edge)
             _edge = edge;
         }
 
+        /// <summary>
+        /// Set the logging level
+        /// </summary>
+        /// <param name="loglevel">log level</param>
+        public static void SetLogLevel(string loglevel)
+        {
+            if (_restClient != null)
```
Comment (childish-sambino) [inline src/Twilio/Twilio.cs:136]: Should really be `if (_logLevel != logLevel)`
